feat(a380x/fctl): Automatic THS trim after engine start or touch and go

flogross89 opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
Adds auto THS trim after engine start or when performing touch and go.
Also, cleans up some XML for the pedestal and the visibility conditions for the magenta GW CG box in the PFD's pitch trim indicator. 

Conditions for engine start: 

- First engine is started and a hydraulic system (G or Y) is pressurized OR
- Ground spoilers become armed

Conditions for touch and go: 

- On ground for at least one second
- Airspeed less than 80kts
- One of the following actions performed: Flaps lever moved, ground spoilers disarmed or at least three thrust levers moved out of IDLE

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
1. Start cold&dark at gate
2. Perform engine start per SOP, THS trim should move automatically
3. Perform take-off, traffic pattern and touch and go. THS trim should move automatically during touch and go
Verify, that magenta PFD pitch trim indicator is only visible after first engine started and HYD is pressurized, and disappears after takeoff (not visible for landing) 
Try to find reference videos to verify whether the trim speed is comparable. Look at the degrees per second, not the overall time. 

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo**, **flybywire-aircraft-a380-842 (4K)** or **flybywire-aircraft-a380-842 (8K)** download link at the bottom of the page

aadee9940 commented:
I know the code is still ready for review, but i had a bit of time so just wanted to try it out. 

Engine start automatic THS:
https://github.com/user-attachments/assets/240ac7c7-6321-4efe-bfe5-1ff833cdf886

Touch and go:
This one I am not sure about, after doing the traffic pattern, on the approach, shouldn't LAND and FLARE be on the FMA? As well as ROLLOUT. Anyways, during touch and go, set flaps to 4->3 and TOGA. The THS value on the SD was changing. 

https://github.com/user-attachments/assets/df1703c0-91da-49ae-9973-c84e27033545

Ueeew commented:
@flogross89 I just wanted to give a few tips or improvements. The first thing I noticed is that process of setting the trim automatically takes way to long. In the real aircraft the process of setting the THS takes only about 2-3 seconds(Real video for reference). In the FBW it takes about 5 or even longer. 
The second thing is that the purple box, indicating the optimal THS setting ,is IR only shown a few seconds prior to auto THS setting process (shown in video). In FBW the box is shown the whole time as long as THS setting is visible on PFD. 
https://m.youtube.com/watch?v=uEgw2p8J0ZA&t=775s&pp=2AGHBpACAcoFDUEzODAgdGhzIHRyaW0%3D 
The engine stat sequence is at minute 2:45. Unfortunately you may have to zoom a bit in to see the process. 
hopefully this helps

flogross89 commented:
@aadee9940 Thanks a lot! Have you considered joining our QA team? 
It seems you didn’t arm any approach mode for your landing, but instead followed the flight plan guidance down to the runway. Did you tune an ILS, and press APP to intercept LOC and G/S? 
The FMA modes are untouched by this PR, so that also wouldn’t be related.

flogross89 commented:
@Ueeew Very helpful, thanks! Indeed, the speed is far too slow, might need to check my triggering mechanism again, I believe I have to use another method. 
about the magenta box: Also well spotted, I suppose the GW CG comes from the FQMS, which only sends these values after engine start or some other condition. I‘ll investigate in my references.

aadee9940 commented:
@flogross89 unfortunately I don't have access to the FBW server as before, but if the mods want to have a look at my appeal LMK 😊

Yeah it completely left my mind I have to program my automatics to get proper FMA modes on landing, my fault 😂.
It was a fully visual circuit so yeah, but I hope my test shows what you are looking for? (The THS trim moving on the SD)
Wasn't really sure if this was EXACTLY what you were looking for..

flogross89 commented:
@aadee9940 Ah, didn't know that. I'm not a mod so I don't know any details, but I trust their process. Yep, that's what I wanted to see, 

@Ueeew I changed the trim triggering mechanism, it should now use the full available speed (which is limited by the PRIMs, so that should be accurate).

Ueeew commented:
@flogross89 could you just merge the master again into this PR so that it’s up-to-Date. Thanks

flogross89 commented on fbw-a380x/src/systems/instruments/src/PFD/PitchTrimDisplay.tsx:
```diff
@@ -33,15 +35,32 @@ export class PitchTrimDisplay extends DisplayComponent<{ bus: EventBus; visible:
+  private readonly greenHydPressurized = ConsumerSubject.create(this.sub.on('hydGreenSysPressurized'), false);
+  private readonly yellowHydPressurized = ConsumerSubject.create(this.sub.on('hydYellowSysPressurized'), false);
+  private readonly oneHydPressurized = MappedSubject.create(
+    ([g, y]) => g || y,
```
Thanks for the review! Changed.

Fowl38 commented:
> @flogross89 I just wanted to give a few tips or improvements. The first thing I noticed is that process of setting the trim automatically takes way to long. In the real aircraft the process of setting the THS takes only about 2-3 seconds(Real video for reference). In the FBW it takes about 5 or even longer. The second thing is that the purple box, indicating the optimal THS setting ,is IR only shown a few seconds prior to auto THS setting process (shown in video). In FBW the box is shown the whole time as long as THS setting is visible on PFD. https://m.youtube.com/watch?v=uEgw2p8J0ZA&t=775s&pp=2AGHBpACAcoFDUEzODAgdGhzIHRyaW0%3D The engine stat sequence is at minute 2:45. Unfortunately you may have to zoom a bit in to see the process. hopefully this helps

We can see on your video, on the F/O's PFD that the THS took 11 seconds to be set (from 2min55 to 3min06 of the video).

flogross89 commented:
> After landing, should the THS FOR value move to the current GW CG value?
> 
> ![image](https://private-user-images.githubusercontent.com/46121009/420937044-0b49c420-7528-414e-b4b2-f9fae4761f4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE2MTYwMjEsIm5iZiI6MTc0MTYxNTcyMSwicGF0aCI6Ii80NjEyMTAwOS80MjA5MzcwNDQtMGI0OWM0MjAtNzUyOC00MTRlLWI0YjItZjlmYWU0NzYxZjRhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzEwVDE0MDg0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlMjhkNTg5NWI0Y2NhMTE0ZDBjYzdhZmM0NDU0ZDk2ZjQ5M2ZlNTQ1OGMwYTIxMzliZDQ4ZWYyZDAxZGU0NWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yKIvUFHVjVGHI4zhsTM0IyCmjHmj1CFyIJ8JisLV0PE)

After landing, it should move to the maintenance position (zero degrees, neutral). This also happened before the PR since it’s the same behavior as on the a320s. 

In case of go-around however, it will move to the GW CG position again

heclak commented:
A few notes from an initial test.

1. Magenta line appears on the gauge once the display powers on.
2. The pink THS value only shows up a few moments after the trimming is completed. Based on the video, the THS value and magenta line should appear at the same time.

Can you confirm how should the THS trim behaviour be observed during a touch and go since the THS gauge doesn't appear during the touch and go? Should the pitch trim position be noted prior to takeoff and the pitch trim should return to the same pitch trim value during the touch and go?

flogross89 commented:
@heclak Thanks for testing!
1. Do you mean just the magenta line was visible, but nothing else? I discovered an issue with visible vs. inherit CSS properties, that should/could fix that. Didn't have time to check it in sim, sorry.
2. Didn't fully understand that, do you mean the automatic THS starts moving before the magenta GW CG values are being displayed? There could be one explanation for this: The Auto THS starts as soon as one HYD system is powered, the GW CG is displayed as soon as one HYD system is powered AND at least one engine is started (N1 > 15%). I believe our HYD systems are powered very early on, I don't know whether the engine-driven HYD pumps are completely accurate in that sense. If it's only a few seconds I'd disregard it for now, and check the HYD system in the future. 

When the following conditions apply, the indicator should be shown during touch&go: 
![image](https://github.com/user-attachments/assets/d0e947fd-c367-47fb-9787-12a695c3219a)
Did those conditions apply? 

During touch&go, the pitch trim should be set to the optimal value w.r.t. the current GW CG, measured/computed by the FQMS.

heclak commented:
This is how the gauge appears immediately after power on. I believe the magenta box/lines shouldn't be there?

![image](https://github.com/user-attachments/assets/5facea2e-98ec-412d-babf-bce098008a08)

Once the HYD is powered, the system starts trimming but the trim value that's usually on the right isn't displayed. It trims with the same visual state as above. And the magenta value on the right only appears AFTER the trimming is completed. Refer to the magenta numbers in the screenshot below. Based on the rough footage of the auto THS working in the reference video above, the magenta markings in the middle of the trim gauge and the THS target on the right should both appear at the same time once the HYD is powered?

![image](https://github.com/user-attachments/assets/90a1bed0-b86a-4eb3-ace6-9e94fe940505)

I'll have to give the touch and go a try again. The indicator didn't appear for me, maybe one of the conditions wasn't fulfilled on my run.

heclak commented:
I think the color of the text might need adjustment too? Looking through this video, maybe the text is white before the magenta lines appear? 
![image](https://github.com/user-attachments/assets/a4eb6f07-06db-44ee-a00b-6edfd5db8ada)

https://youtu.be/p28erS6uWHE?si=PzILc6PvMB_cIwpD&t=1423

flogross89 commented:
@heclak Ah, now I see it. Thanks for spotting that, I missed that nuance before. Now:
Before engine start: 
![image](https://github.com/user-attachments/assets/329860f1-8d22-402a-9887-06ab92e06a65)

After engine start:
![image](https://github.com/user-attachments/assets/f0770ffe-88f8-4104-b285-b224ccf8af31)

About the difference in timing of auto trim vs. PFD indications: I changed the conditions, the cyan indications now appear before the auto trim moves. It's still not 100% synced, that would increase the complexity in the PFD indication unnecessarily, let's wait for the PRIM implementation for that.

Could you please give it one last try?

flogross89 commented:
@heclak Thanks for trying it.
What I can see in the video:
1) PFD pitch trim indicator is shown below 50ft during landing, that's wrong. I added the "on Ground" condition (without confirmation time, as per FCOM) for that case, should only be displayed on ground when at least three thrust levers are moved out of idle

Concerning your questions: 
1) Triggering of auto trim: From what I can see, it behaves correctly, doesn't it? As soon as you moved the thrust levers out of idle, the auto trim starts. 
However, the magenta box in the PFD pitch trim indicator is displayed too late, indeed I waited for a FWC flight phase change. I softened that, so it should display the magenta box faster. Hopefully it didn't worsen any other situations
2) TLA is the thrust lever angle, i.e. doesn't correspond to thrust or N1, but only to the angle of the physical lever (see doc for `A32NX_AUTOTHRUST_TLA:{index}`). I updated the condition to check all thrust levers

heclak commented:
@flogross89 

Thanks for the changes!

Good catch on the trim moving first. It was such a small trim change that I missed it even with multiple viewings of the video. I need fresher eyes. 

I'll try and keep a look out for other conditions that might be affected by the soften of the magenta box appearance. Probably need to test the post landing appearance too.

flogross89 commented:
@heclak Thanks again, this helped my find two other issues: 
1) Somehow the SimVarPublisher wasn't added for thrust levers 3&4 (one of those a32nx copy&paste errors), so the idle thrust check didn't trigger. This fixes the issue when only moving the thrust levers and not touching the ground spoilers or flaps
2) The optimal pitch trim was set to 0 once we landed, which doesn't make sense (even not below 30kts, since it's not the optimal position, but the maintenance position. It will be displayed in white anyways when it returns to 0 after landing. 

Just pushed the changes. I feel we're converging on a final solution here :)

heclak commented:
<!-- !! DO NOT REMOVE THIS LINE !! - QUALITY ASSURANCE REPORT -->
<!-- Refer to the https://docs.flybywiresim.com/dev-corner/qa-process/#reporting-on-tests for details about this template -->

<!-- Please use one or the other depending on your Discord Role -->
Quality Assurance Tester/Trainee Report

Discord Username  : Heclak <!-- Use your username, not your display name! -->
Object of testing : #9843 <!-- PR Number -->
Aircraft          : A380X <!-- Choose only one -->
Tier of Testing   : 1
Date              : 27/03/2025 <!-- Date of test in DD/MM/YYYY format -->

Testing Process: 
<!-- List the process - Required section --> 

1. Start Cold and Dark and startup aircraft as per SOP
2. Perform engine start per SOP, THS trim should move automatically ✔️ 
3. Perform take-off, traffic pattern and touch and go. THS trim should move automatically during touch and go
    - Touch and Go 1 - Moved flaps. Check THS responded correctly ✔️ 
    - Touch and Go 2 - Moved thrust levers. Check THS responded correctly ✔️ 
    - Touch and Go 3 - Disarm spoilers. Check THS responded correctly ✔️ 
4. Full stop landing. Check pitch trim position indication appears under 30 knots and THS trims to 0 position.

Testing Results: 
Passed <!-- Required section --> 

Negatives:       
<!-- Ommit this section if none are found -->
Noticed two minor details that I think are not critical but doesn't need to be in the scope of this PR.

1. In this specific test, I noticed there is a rounding difference between the left and right values. In the EFB, the GWCG is 34.76%. Which makes me suspect it's a rounding method difference between the two values.

![image](https://github.com/user-attachments/assets/33ae5dbd-8de3-4a9d-b3cd-962420c12b46)

2. The movement speed of the striped portion of the pitch trim position indicator might need tweaking. It currently looks to move in an almost 1:1 ratio with the left side of the trim indicator. However, in the video, it appears it should be moving much faster with a per degree change in THS. This is purely visual. https://youtu.be/uEgw2p8J0ZA?si=yRknN00h7iLLbLUO&t=173

It think both issues could be handled in a separate PR if needed.

Conclusions: 
<!-- Ommit section if nothing to add -->

Core function of the THS is functioning as designed and no blocking issues. The two minor negatives could be handled separately if needed.

Media:
<!-- Ommit this section if no media to add -->
